src/dmd/mtype.d

```diff
@@ -601,77 +612,81 @@ extern (C++) abstract class Type : ASTNode
         {
             if (t1.parameterList.length != t2.parameterList.length)
-                goto Ldistinct;
+                return distinct();
 
             foreach (i, fparam1; t1.parameterList)
             {
                 Parameter fparam2 = t2.parameterList[i];
+                void checkParametersCovariant() { notcovariant |= !fparam1.isCovariant(t1.isref, fparam2); }
+                if (fparam1.type.equals(fparam2.type))
+                {
+                    checkParametersCovariant();
+                    continue;
+                }
+                Type tp1 = fparam1.type;
+                Type tp2 = fparam2.type;
+                if (tp1.ty != tp2.ty)
+                    return distinct();
 
-                if (!fparam1.type.equals(fparam2.type))
+                if (auto tc1 = tp1.isTypeClass())
                 {
-                    Type tp1 = fparam1.type;
-                    Type tp2 = fparam2.type;
-                    if (tp1.ty == tp2.ty)
-                    {
-                        if (auto tc1 = tp1.isTypeClass())
-                        {
-                            if (tc1.sym == (cast(TypeClass)tp2).sym && MODimplicitConv(tp2.mod, tp1.mod))
-                                goto Lcov;
-                        }
-                        else if (auto ts1 = tp1.isTypeStruct())
-                        {
-                            if (ts1.sym == (cast(TypeStruct)tp2).sym && MODimplicitConv(tp2.mod, tp1.mod))
-                                goto Lcov;
-                        }
-                        else if (tp1.ty == Tpointer)
-                        {
-                            if (tp2.implicitConvTo(tp1))
-                                goto Lcov;
-                        }
-                        else if (tp1.ty == Tarray)
-                        {
-                            if (tp2.implicitConvTo(tp1))
-                                goto Lcov;
-                        }
-                        else if (tp1.ty == Tdelegate)
-                        {
-                            if (tp1.implicitConvTo(tp2))
-                                goto Lcov;
-                        }
-                    }
-                    goto Ldistinct;
+                    if (tc1.sym == (cast(TypeClass)tp2).sym && MODimplicitConv(tp2.mod, tp1.mod))
+                        checkParametersCovariant();
+                    else return distinct();
+                }
+                else if (auto ts1 = tp1.isTypeStruct())
+                {
+                    if (ts1.sym == (cast(TypeStruct)tp2).sym && MODimplicitConv(tp2.mod, tp1.mod))
+                        checkParametersCovariant();
+                    else return distinct();
+                }
+                else if (tp1.ty == Tpointer)
+                {
+                    if (tp2.implicitConvTo(tp1))
+                        checkParametersCovariant();
+                    else return distinct();
```

Comment:
Actually, couldn't these conditions be inverted, then you can have the checkParametersCovariant at the end of the scope unconditionally?


```suggestion
                    if (!tp2.implicitConvTo(tp1))
                        return distinct();
```
(then the other call goes...)